ENG-1946 - feat: introduce React package documentation (#143)
Basis-Theory/docs: source/includes/elements/react/_basis_theory_provider.md

# BasisTheoryProvider

```jsx
import { 
  useBasisTheory, 
  BasisTheoryProvider
} from '@basis-theory/basis-theory-react';

> import { TextElement } from '@basis-theory/basis-theory-react/elements';

Reviewer: If the package contains non-elements stuff too, I'm fine with importing from a subdirectory to not pollute the global index file with elements-related things too. I've several larger packages require you to import this way.

> import { TextElement } from '@basis-theory/basis-theory-react/elements';

Reviewer 2: Does this have other benefits like tree-shaking? i'm personally good with either, my IDE does that for me anyway. 🤣

> import { TextElement } from '@basis-theory/basis-theory-react/elements';

Author: Yes.

> import { TextElement } from '@basis-theory/basis-theory-react/elements';

Author: Per external feedback, added root import. Doesn't impact tree shaking.

const App = () => {
  // creates a new instance of BasisTheory class

>   const { bt } = useBasisTheory('test_1234567890', { elements: true });

Reviewer: Not sure what the right terminology is, but instance of the `BasisTheory class` sounds strange here, maybe something more like
// creates a new instance of the BasisTheory [client|sdk|library]

Reviewer 2: maybe even just "creates a new instance of the BasisTheory" since it is already in that context?

Author: `BasisTheory` is indeed a class and the way the instance is handled is important. Throughout the documentation we refer to "BasisTheory instance". I am okay changing all references to a more descriptive or understandable terminology.

  
  if (bt) {
    // able to call BasisTheory methods
  }
  
  return <BasisTheoryProvider bt={bt}>
      <MyComponent />
    </BasisTheoryProvider>;
}

const MyComponent = () => {
  // calling this hook with no attributes grabs the BasisTheory instance from Context  
  const { bt } = useBasisTheory();

  if (bt) {
    // able to call BasisTheory methods
  }
  
  return <TextElement /> ; // Element will also grab it from the Context
}
```

This <a href="https://reactjs.org/docs/context.html" target="_blank">Context Provider</a> shares a `BasisTheory` instance to your component tree, making it available for Basis Theory Elements or other custom components. 

Property   | Required | Type                | Description
---------- | -------- | ------------------- | -----------
`bt`       | false    | *BasisTheoryReact*  | [Instance](#basistheoryreact) to be provided in the context. <br><i>Note: this is not required because initialization happens asynchronously. See [`useBasisTheory`](#usebasistheory).</i>

<aside class="notice">

>   <span>Elements will prioritize the <code>BasisTheory</code> instance passed to it's own <code>bt</code> property over an instance passed in to a parent <code>BasisTheoryProvider</code>.</span>

Reviewer: This is a little confusing to me - is this talking about what would happen if you passed another `bt` instance into `<TextElement bt={bt} />`? And why would I ever do that - would could I gain from knowing that exists?

Author: `BasisTheoryProvider` usage is optional. If you have a more localized use case, where you just want to collect, say a Card, in your whole application, you can just go for `useBasisTheory` + `CardElement`.
Maybe we need to do a better job explaining that. Did you have the chance to go through the running documentation?

# BasisTheoryProvider

```jsx
import { 
  useBasisTheory, 
  BasisTheoryProvider
} from '@basis-theory/basis-theory-react';
import { TextElement } from '@basis-theory/basis-theory-react/elements';

const App = () => {
  // creates a new instance of BasisTheory class
  const { bt } = useBasisTheory('test_1234567890', { elements: true });
  
  // bt variable stays undefined until properly initialized
  if (bt) {
    // able to call BasisTheory methods
  }
  
  return <BasisTheoryProvider bt={bt}>
      <MyComponent />
    </BasisTheoryProvider>;
}

const MyComponent = () => {
  // calling this hook with no attributes grabs the BasisTheory instance from Context  
  const { bt } = useBasisTheory();

  if (bt) {
    // able to call BasisTheory methods
  }
  
  return <TextElement /> ; // Element will also grab it from the Context
}
```

> This <a href="https://reactjs.org/docs/context.html" target="_blank">Context Provider</a> is useful for sharing a `BasisTheory` instance in your component tree, making it available for Basis Theory Elements or custom components. 

Reviewer 2: ```suggestion
This <a href="https://reactjs.org/docs/context.html" target="_blank">Context Provider</a> is shares a `BasisTheory` instance to your component tree, making it available for Basis Theory Elements or other custom components. 
```

Property   | Required | Type           | Description
---------- | -------- | -------------- | -----------
`bt`       | false    | *BasisTheory*  | Instance to be provided in the context. <br><i>Note: this is not required because initialization happens asynchronously. See [`useBasisTheory`](#usebasistheory).</i>

<aside class="notice">

>   <span>Elements can consume the <code>BasisTheory</code> instance from Context if nested under <code>BasisTheoryProvider</code>, but will prioritize their <code>bt</code> prop over it.</span>

Reviewer 2: ```suggestion
  <span>Elements will prioritize the <code>BasisTheory</code> instance passed to it's own <code>bt</code> property over and instance passed in to a parent <code>BasisTheoryProvider</code>.</span>
```
